workflows/prognostic_c48_run/runtime/monitor.py

```diff
@@ -0,0 +1,107 @@
+            variables: list of variables with names like
+                `tendency_of_{variable}_due_to_{name}`. Used to infer the variables
+                to be monitored.
+            state: The mutable object to monitor
+            timestep: the length of the timestep used to compute the tendency
+        """
+        # need to consume variables into set to use more than once
+        var_set = set(variables)
+        self.tendency_variables = filter_tendency(var_set)
+        self.storage_variables = filter_storage(var_set)
+        self._state = state
+        self.timestep = timestep
+
+    def __call__(
+        self, name: str, func: Callable[[], Diagnostics],
+    ) -> Callable[[], Diagnostics]:
+        """Decorator to add tendency monitoring to an update function
+
+        This will add the following diagnostics:
+        - `tendency_of_{variable}_due_to_{name}`
+        - `storage_of_{variable}_path_due_to_{name}`. A pressure-integrated version
+        of the above
+        - `storage_of_mass_due_to_{name}`, the total mass tendency in Pa/s.
+
+        Args:
+            name: the name to tag the tendency diagnostics with
+            func: a stepping function
+        Returns:
+            monitored function. Same as func, but with tendency and mass change
+            diagnostics inserted in place
+        """
+
+        def step() -> Diagnostics:
+
+            vars_ = self.storage_variables | self.tendency_variables
+            delp_before = self._state[DELP]
+            before = {key: self._state[key] for key in vars_}
+            diags = func()
+            delp_after = self._state[DELP]
+            after = {key: self._state[key] for key in vars_}
+
+            for variable in self.storage_variables:
+                path_before = vcm.mass_integrate(before[variable], delp_before, "z")
+                path_after = vcm.mass_integrate(after[variable], delp_after, "z")
+
+                diag_name = f"storage_of_{variable}_path_due_to_{name}"
+                diags[diag_name] = (path_after - path_before) / self.timestep
+                if "units" in before[variable].attrs:
+                    diags[diag_name].attrs["units"] = (
+                        before[variable].units + " kg/m**2/s"
+                    )
+
+            # Compute statistics
+            for variable in self.tendency_variables:
+                diag_name = f"tendency_of_{variable}_due_to_{name}"
+                diags[diag_name] = (after[variable] - before[variable]) / self.timestep
+                if "units" in before[variable].attrs:
+                    diags[diag_name].attrs["units"] = before[variable].units + "/s"
+
+            mass_change = (delp_after - delp_before).sum("z") / self.timestep
+            mass_change.attrs["units"] = "Pa/s"
+            diags[f"storage_of_mass_due_to_{name}"] = mass_change
+            return diags
+
+        # ensure monitored function has same name as original
+        step.__name__ = func.__name__
```

Comment:
pre-existing: `functools.wraps` handles copying over the docstring as well, and might be a better high-level function to use here.

Reply:
There is no coverage of the clock timing code, so I don't want to risk breaking it.